infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java

```diff
@@ -248,10 +257,12 @@ private TxResolutionResponse txnCanCommit(TxResolutionInfo txInfo) {
 
         // A transaction can start with a timestamp issued from a previous
         // epoch, so we need to reject transactions that have a snapshot
-        // timestamp with a different epoch than the this sequencer's epoch.
-        if (txSnapshotTimestamp.getEpoch() != sequencerEpoch) {
-            log.debug("ABORT[{}] snapshot-ts[{}] current epoch[{}]", txInfo,
-                    txSnapshotTimestamp, sequencerEpoch);
+        // timestamp less than the firstConsecutiveEpoch since we are sure
+        // this sequencer is always the primary sequencer after this epoch.
+        long tokenEpoch = txSnapshotTimestamp.getEpoch();
+        if (tokenEpoch != sequencerEpoch && tokenEpoch < firstConsecutiveEpoch) {
```

Comment:
In what case will you end up tokenEpoch > firstConsecutiveEpoch? From what I understand, you will end up with either:

1) tokenEpoch < firstConsecutiveEpoch
2) tokenEpoch == firstConsecutiveEpoch

Reply:
Every time system epoch changes, the primary sequencer will get a bootstrap request. In the bootstrap handling logic, firstConsecutiveEpoch will be updated if the bootstrap epoch is a consecutive epoch to the current sequencer epoch:https://github.com/CorfuDB/CorfuDB/blob/17f8d5314d0b415a9127f48333d1605580d3b79e/infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java#L449

In this way, we keep a consecutive range: [firstConsecutiveEpoch, sequencerEpoch]. This sequencer is the primary sequencer for every epoch within this range. So when a transaction has a tokenEpoch within this range, it could be accepted. In this case, from the time when the transaction starts to the time when transaction wants to commit, the primary sequencer never changes (this is the case when we end up tokenEpoch >= firstConsecutiveEpoch).

Reply:
I see. I though you were updating the `firstConsecutiveEpoch` when `bootstrapMsgEpoch == sequencerEpoch + 1`. I would be nice if you expressed this logic in terms of a sliding window/range:

`strictlyMonotonicEpochWindow = [firstConsecutiveEpoch, sequencerEpoch]`

.. and then do the boundary check in form of 'range.contains(tokenEpoch)'. You might want to take a look at: https://google.github.io/guava/releases/19.0/api/docs/com/google/common/collect/Range.html

Reply:
Actually my first thought was also using Range, but then I found the approach I'm currently using is more efficient since it only keeps a `long` variable and does a simple long comparison, instead of using `contains`, which is more costly.

Reply:
You really should not worry about performance in this case, since epoch changes happen very infrequently (relative to other events in the system). I would argue that this is a premature optimization.

Readability and understandability are way more important.

Reply:
Yes I agree, will do.

Reply:
Also, you could extract `[firstConsecutiveEpoch, sequencerEpoch]` into a class (it could be ConsequtiveEpoch?) also move all related logic to that class and cover the class by unit tests.

Reply:
@vkeko Add a helper method for range check as discussed.